### PR TITLE
issue-108: Add saved search feature

### PR DESCRIPTION
**Pull request by lipim12**
Closing https://github.com/attivio/suit/pull/169 and making a new PR from a branch on suit (and not my forked repo)

**Addressing requested changes on #110**

_A few notes about refactoring with `MetadataManager` api class:_

- It is made for handling metadata management. You can think of a few examples where a component needs to make an API request but it is not an authentic search to be put in searcher or search. You can add such functions in` MetadataManager`.

- Originally this idea about refactoring with MetadataManager has been discussed before with Nikhil, Kristi, Luis, and Brandon.

- Workflow, Zone, Table for Saved Searches have been hardcoded as of now, with an assumption that `.metadata` zone would become OOTB in upcoming releases.

_How to use `MetadataManager`:_

- Create an instance of `MetadataManager` in the constructor of a component where you need to use its services. Initialize it with Searcher.Search.

```
example: MetadataManager;

constructor(props: ThisComponentProps) {
  super(props);
  this.example = new MetadataManager(this.context.searcher.search);
}

```
- To use a function of `MetadataManager`, call it using the created instance of `MetadataManager` and pass in required arguments.

```
sampleFunction = () => {
  ...
  ...
  this.example.someFunction(arg0, arg1);
  ...
}

```

### CODE REVIEW

File: src/components/ListSearchResult.js [hunk 1]
```diff
@@ -77,22 +82,8 @@ class InnerListSearchResult extends React.Component<InnerListSearchResultDefault
...
-      <ListSearchResult
-        document={doc}
-        position={position}
-        baseUri={baseUri}
-        key={key}
-        hide360Link={hide360Link}
-      />
-    );
+  static renderer(doc: SearchDocument, position: number, baseUri: string, key: string, hide360Link: boolean = false) {
+    return <ListSearchResult document={doc} position={position} baseUri={baseUri} key={key} hide360Link={hide360Link} />;
```
Comment (klk1010) [inline src/components/ListSearchResult.js]: Prettier? Let's keep this as it was.
```suggestion
static renderer(
  doc: SearchDocument,
  position: number,
  baseUri: string,
  key: string,
  hide360Link: boolean = false,
  ) {
    return (
      <ListSearchResult
        document={doc}
        position={position}
        baseUri={baseUri}
        key={key}
        hide360Link={hide360Link}
      />
    );
```

File: src/components/ListSearchResult.js [hunk 2]
```diff
@@ -157,52 +148,26 @@ class InnerListSearchResult extends React.Component<InnerListSearchResultDefault
         const docResults = tableDocs.map((tableDoc, index) => {
           const childPosition = index + 1;
           return (
-            <ListSearchResult
-              document={tableDoc}
-              key={tableDoc.getFirstValue('.id')}
-              position={childPosition}
-              baseUri={baseUri}
-            />
+            <ListSearchResult document={tableDoc} key={tableDoc.getFirstValue('.id')} position={childPosition} baseUri={baseUri} />
```
Comment (klk1010) [inline src/components/ListSearchResult.js]: Same
```suggestion
            <ListSearchResult
              document={tableDoc}
              key={tableDoc.getFirstValue('.id')}
              position={childPosition}
              baseUri={baseUri}
            />
```

File: src/components/ListSearchResult.js [hunk 3]
```diff
@@ -157,52 +148,26 @@ class InnerListSearchResult extends React.Component<InnerListSearchResultDefault
...
           );
         });
-        const docResultsList = (
-          <div className="attivio-nested-search-results">
-            {docResults}
-          </div>
-        );
+        const docResultsList = <div className="attivio-nested-search-results">{docResults}</div>;
```
Comment (klk1010) [inline src/components/ListSearchResult.js]: same

File: src/components/ListSearchResult.js [hunk 4]
```diff
@@ -157,52 +148,26 @@ class InnerListSearchResult extends React.Component<InnerListSearchResultDefault
...
         tabInfos.push(new TabInfo(label, tabTable, docResultsList));
       });
 
-      const tabLabel = doc.children.length === 1
-        ? 'One Child Record:'
-        : `${doc.children.length} Child Records:`;
+      const tabLabel = doc.children.length === 1 ? 'One Child Record:' : `${doc.children.length} Child Records:`;
```
Comment (klk1010) [inline src/components/ListSearchResult.js]: same

File: src/components/ListSearchResult.js [hunk 5]
```diff
@@ -157,52 +148,26 @@ class InnerListSearchResult extends React.Component<InnerListSearchResultDefault
...
 
       return (
-        <TabPanel
-          tabInfos={tabInfos}
-          activeTabId={this.state.currentTab}
-          tabChanged={this.tabChanged}
-          tabLabel={tabLabel}
-          nested
-        />
+        <TabPanel tabInfos={tabInfos} activeTabId={this.state.currentTab} tabChanged={this.tabChanged} tabLabel={tabLabel} nested />
```
Comment (klk1010) [inline src/components/ListSearchResult.js]: same

File: src/components/ListSearchResult.js [hunk 6]
```diff
@@ -157,52 +148,26 @@ class InnerListSearchResult extends React.Component<InnerListSearchResultDefault
...
-      baseUri,
-      document: doc,
-      entityFields,
-      position,
-      showRatings,
-      showScores,
-      showTags,
-      hide360Link,
-    } = this.props;
+    const { baseUri, document: doc, entityFields, position, showRatings, showScores, showTags, hide360Link } = this.props;
```
Comment (klk1010) [inline src/components/ListSearchResult.js]: same

File: src/components/ListSearchResult.js [hunk 7]
```diff
@@ -215,30 +180,26 @@ class InnerListSearchResult extends React.Component<InnerListSearchResultDefault
     const moreLikeThisQuery = doc.getFirstValue('morelikethisquery');
     const docTags = doc.getAllValues('tags');
 
-    const userDefinedEntityFieldsExist = configuration
-      && configuration.state
-      && configuration.state.ALL
-      && configuration.state.ALL.entityFields;
+    const userDefinedEntityFieldsExist =
```
Comment (klk1010) [inline src/components/ListSearchResult.js]: prettier

File: src/components/ListSearchResult.js [hunk 8]
```diff
@@ -215,30 +180,26 @@ class InnerListSearchResult extends React.Component<InnerListSearchResultDefault
...
+      configuration && configuration.state && configuration.state.ALL && configuration.state.ALL.entityFields;
 
-    const userEntityFields = userDefinedEntityFieldsExist
-      ? configuration.state.ALL.entityFields
-      : entityFields;
+    const userEntityFields = userDefinedEntityFieldsExist ? configuration.state.ALL.entityFields : entityFields;
```
Comment (klk1010) [inline src/components/ListSearchResult.js]: prettier

File: src/components/ListSearchResult.js [hunk 9]
```diff
@@ -215,30 +180,26 @@ class InnerListSearchResult extends React.Component<InnerListSearchResultDefault
...
     return (
       <div className=" attivio-search-result">
         <div className="attivio-search-result-col">
           <DocumentType docType={table} position={position} />
-          <DocumentThumbnail
-            uri={thumbnailUri}
-            previewUris={previewUri}
-            previewTitle={doc.getFirstValue(FieldNames.TITLE)}
-          />
+          <DocumentThumbnail uri={thumbnailUri} previewUris={previewUri} previewTitle={doc.getFirstValue(FieldNames.TITLE)} />
```
Comment (klk1010) [inline src/components/ListSearchResult.js]: prettier

File: src/components/ListSearchResult.js [hunk 10]
```diff
@@ -215,30 +180,26 @@ class InnerListSearchResult extends React.Component<InnerListSearchResultDefault
...
           <dl className="attivio-labeldata-stacked attivio-labeldata-stacked-search-results">
             {showRatings && (
               <div>
                 <dt>Rating</dt>
                 <dd>
-                  <StarRating onRated={(rating) => { this.rateDocument(doc, rating); }} />
+                  <StarRating
+                    onRated={(rating) => {
```
Comment (klk1010) [inline src/components/ListSearchResult.js]: prettier

File: src/components/ListSearchResult.js [hunk 11]
```diff
@@ -256,12 +217,7 @@ class InnerListSearchResult extends React.Component<InnerListSearchResultDefault
             <Col xs={7} sm={7}>
               <SearchResultBody body={text} />
               {showTags && (
-                <SearchResultTags
-                  tags={docTags}
-                  moreLikeThisQuery={moreLikeThisQuery}
-                  docId={docId}
-                  hide360Link={hide360Link}
-                />
+                <SearchResultTags tags={docTags} moreLikeThisQuery={moreLikeThisQuery} docId={docId} hide360Link={hide360Link} />
```
Comment (klk1010) [inline src/components/ListSearchResult.js]: prettier

File: src/components/MetadataManager.js [hunk 12]
```diff
@@ -0,0 +1,224 @@
...
+    const mode = 'DELETE';
+    this.updateSavedSearches(ss, this.getSavedSearches, mode);
+  }
+
+  updateSavedSearches(ss: SavedSearch, callback: () => void, mode: string = 'ADD'): Promise<any> {
+    return new Promise((resolve, reject) => {
+      // Get session
+      const baseUri = this.context.searcher.search.baseUri;
+      const connectUri = `${baseUri}/rest/ingestApi/connect`;
+      fetch(connectUri, { credentials: 'include' })
```
Comment (klk1010) [inline src/components/MetadataManager.js]: This logic should be in the `api` directory instead of inside a component. Avoid putting complex fetching logic in components.

File: src/components/SavedSearchRenderer.js [hunk 13]
```diff
@@ -0,0 +1,253 @@
...
+    (this: any).showSaveSearchModal = this.showSaveSearchModal.bind(this);
+    (this: any).hideSaveSearchModal = this.hideSaveSearchModal.bind(this);
+    (this: any).searchTitleEntered = this.searchTitleEntered.bind(this);
+    (this: any).renderSavedQuery = this.renderSavedQuery.bind(this);
+    (this: any).renderSavedQueryHeader = this.renderSavedQueryHeader.bind(this);
+  }
+
+  state: SavedSearchRendererState;
+
+  componentWillMount() {
```
Comment (klk1010) [inline src/components/SavedSearchRenderer.js]: Avoid `componentWillMount`. Use `componentDidMount` instead.

File: src/components/SavedSearchRenderer.js [hunk 14]
```diff
@@ -0,0 +1,253 @@
...
+    this.context.metadataManager.getSavedSearches();
+    this.setState({
+      savedSearchList: this.context.metadataManager.state.savedSearchList,
+    });
+  }
+
+  // applies the saved search which includes a query and facet filters
+  applySavedSearch(query: string) {
+    const { location } = this.props;
```
Comment (klk1010) [inline src/components/SavedSearchRenderer.js]:
```suggestion
    const { location, history } = this.props;
    history.push({
```

File: src/components/SavedSearchRenderer.js [hunk 15]
```diff
@@ -0,0 +1,253 @@
...
+
+  componentWillMount() {
+    this.context.metadataManager.getSavedSearches();
+    this.setState({
+      savedSearchList: this.context.metadataManager.state.savedSearchList,
+    });
+  }
+
+  // applies the saved search which includes a query and facet filters
+  applySavedSearch(query: string) {
```
Comment (klk1010) [inline src/components/SavedSearchRenderer.js]: Instead: 
```
applySavedSearch(search: string) {
    const { location: { pathname }, history } = this.props;
    history.push({
      pathname,
      search,
    });
  }
```

File: src/components/SavedSearchRenderer.js [hunk 16]
```diff
@@ -0,0 +1,253 @@
...
+  constructor(props: SavedSearchRendererProps) {
+    super(props);
+    this.state = {
+      response: undefined,
+      error: undefined,
+      showSaveSearchModal: false,
+      savedSearchTitle: '',
+      savedSearchList: [],
+    };
+    (this: any).applySavedSearch = this.applySavedSearch.bind(this);
```
Comment (klk1010) [inline src/components/SavedSearchRenderer.js]: Instead of using bindings, use arrow functions. This is preferred for new components.

File: src/components/SavedSearchRenderer.js [hunk 17]
```diff
@@ -0,0 +1,253 @@
...
+      savedSearchTitle: this.context.searcher.state.query,
+      showSaveSearchModal: true,
+    });
+  }
+
+  renderSavedQueryHeader() {
+    return (
+      <MenuItem
+        key="saveQueryHeader"
+        onSelect={() => {
```
Comment (klk1010) [inline src/components/SavedSearchRenderer.js]: make `showSaveSearchModal() {` this instead: `showSaveSearchModal = () => {`
Then do this:
```suggestion
        onSelect={this.showSaveSearchModal}
```

File: src/components/SavedSearchRenderer.js [hunk 18]
```diff
@@ -0,0 +1,253 @@
...
+> {
+  static defaultProps: SavedSearchRendererDefaultProps = {};
+
+  static displayName = 'SavedSearchRenderer';
+
+  static contextTypes = {
+    searcher: PropTypes.any,
+  };
+
+  constructor(props: SavedSearchRendererProps) {
```
Comment (klk1010) [inline src/components/SavedSearchRenderer.js]: Because state is not based on props, you don't need a constructor. You can set default state below.

File: src/components/SavedSearchRenderer.js [hunk 19]
```diff
@@ -0,0 +1,253 @@
...
+    };
+    (this: any).applySavedSearch = this.applySavedSearch.bind(this);
+    (this: any).showSaveSearchModal = this.showSaveSearchModal.bind(this);
+    (this: any).hideSaveSearchModal = this.hideSaveSearchModal.bind(this);
+    (this: any).searchTitleEntered = this.searchTitleEntered.bind(this);
+    (this: any).renderSavedQuery = this.renderSavedQuery.bind(this);
+    (this: any).renderSavedQueryHeader = this.renderSavedQueryHeader.bind(this);
+  }
+
+  state: SavedSearchRendererState;
```
Comment (klk1010) [inline src/components/SavedSearchRenderer.js]:
```suggestion
  state: SavedSearchRendererState = {
    response: undefined,
    error: undefined,
    showSaveSearchModal: false,
    savedSearchTitle: '',
    savedSearchList: [],
  };
```

File: src/components/SavedSearchRenderer.js [hunk 20]
```diff
@@ -0,0 +1,253 @@
...
+          }}
+        >
+          No Saved Searches
+        </span>
+      </MenuItem>
+    );
+  }
+
+  renderSavedQuery() {
+    const savedQueryDivider = (
```
Comment (klk1010) [inline src/components/SavedSearchRenderer.js]: Move out to separate function `renderSavedQueryDivider()`

File: src/components/SearchBar.js [hunk 21]
```diff
@@ -417,6 +430,7 @@ class SearchBar extends React.Component<SearchBarDefaultProps, SearchBarProps, S
           {suggestionList}
         </div>
         {languageControl}
+        {SavedSearchComp}
```
Comment (klk1010) [inline src/components/SearchBar.js]:
```suggestion
        {this.props.allowSavedSearch && <SavedSearchRenderer />}
```

File: src/components/SearchResults.js [hunk 22]
```diff
@@ -92,11 +98,7 @@ export default class SearchResults extends React.Component<SearchResultsDefaultP
   static displayName = 'SearchResults';
 
   renderResults() {
-    const {
-      baseUri,
-      format,
-      hide360Link,
-    } = this.props;
+    const { baseUri, format, hide360Link } = this.props;
```
Comment (klk1010) [inline src/components/SearchResults.js]: Revert Prettier changes in this file

File: src/components/MetadataManager.js [hunk 23]
```diff
@@ -0,0 +1,224 @@
...
+ */
+class MetadataManager extends React.Component<MetadataManagerDefaultProps, MetadataManagerProps, MetadataManagerState> {
+  static defaultProps: MetadataManagerDefaultProps = {};
+  static childContextTypes = {
+    metadataManager: PropTypes.any,
+  };
+
+  static displayName = 'MetadataManager';
+
+  constructor(props: MetadataManagerProps) {
```
Comment (klk1010) [inline src/components/MetadataManager.js]: Unnecessary constructor. Prefer arrow functions over function bindings. State is not based on props. Set in line 47. Note that if you do not set an initial state value for a state const, that const will be `undefined` by default. Prefer `null` over `undefined`.

File: src/components/SearchResults.js [hunk 24]
```diff
@@ -152,20 +154,14 @@ export default class SearchResults extends React.Component<SearchResultsDefaultP
   }
 
   render() {
-    const {
-      style: baseStyle = {},
-    } = this.props;
+    const { style: baseStyle = {} } = this.props;
```
Comment (klk1010) [inline src/components/SearchResults.js]: Revert Prettier changes in this file.

File: src/components/SavedSearchRenderer.js [hunk 25]
```diff
@@ -0,0 +1,253 @@
...
+      </MenuItem>
+    );
+  }
+
+  renderSavedQueryList() {
+    return this.state.savedSearchList.length > 0 ? (
+      this.state.savedSearchList.map((e) => {
+        return (
+          <MenuItem key={e.id}>
+            <span // eslint-disable-line
```
Comment (klk1010) [inline src/components/SavedSearchRenderer.js]: Why is this here?
`// eslint-disable-line`

File: src/components/SavedSearchRenderer.js [hunk 26]
```diff
@@ -0,0 +1,253 @@
...
+              onClick={() => {
+                this.applySavedSearch(e.query);
+              }}
+              title={e.title}
+            >
+              {e.titleLabel}
+            </span>
+            <span // eslint-disable-line
+              onClick={() => {
```
Comment (klk1010) [inline src/components/SavedSearchRenderer.js]: Add comments to explain use of `// eslint-disable-line`
Shouldn't this be:
`onClick={(e) => this.context.metadataManager.deleteThisSearch(e.id)}`

File: src/components/SavedSearchRenderer.js [hunk 27]
```diff
@@ -0,0 +1,253 @@
...
+        >
+          <b> SAVE SEARCH </b>
+        </span>
+      </MenuItem>
+    );
+  }
+
+  renderSavedQueryList() {
+    return this.state.savedSearchList.length > 0 ? (
+      this.state.savedSearchList.map((e) => {
```
Comment (klk1010) [inline src/components/SavedSearchRenderer.js]: Since this is just a return with any other declarations, you can use the auto-return feature of arrow functions. Just use parentheses with no curly braces and leave out the return. If you are using auto return and only have one param, that param does not have to be in parentheses either.
```
this.state.savedSearchList.map(e => (
  <MenuItem key={e.id}>
    ...
  <MenuItem />
);
```

**Review by klk1010 — CHANGES_REQUESTED** [review]
Address feedback

File: src/components/SavedSearchRenderer.js [hunk 28]
```diff
@@ -0,0 +1,253 @@
...
+    (this: any).renderSavedQuery = this.renderSavedQuery.bind(this);
+    (this: any).renderSavedQueryHeader = this.renderSavedQueryHeader.bind(this);
+  }
+
+  state: SavedSearchRendererState;
+
+  componentWillMount() {
+    this.context.metadataManager.getSavedSearches();
+    this.setState({
+      savedSearchList: this.context.metadataManager.state.savedSearchList,
```
Comment (klk1010) [inline src/components/SavedSearchRenderer.js]: Is there a reason why we're doing a subscription fork here? It's generally better to reference the original data source unless you need a reference to the data as it was at mount time.

**Review by klk1010 — APPROVED** [review]
Nice work!

**Review by luigivaldez — CHANGES_REQUESTED** [review]
See the comments below...

Also, can Nikhil's old PR be closed? https://github.com/attivio/suit/pull/110

**Comment by lipim12** [comment]
Yes you can close Nikhil's PR

File: src/api/MetadataManager.js [hunk 29]
```diff
@@ -0,0 +1,195 @@
...
+import SimpleQueryRequest from '../api/SimpleQueryRequest';
+import SavedSearch from '../api/SavedSearch';
+import AuthUtils from '../util/AuthUtils';
+import Search from '../api/Search';
+
+/**
+ * Class that manages all the metadata management, manages functions that require api calls
+ * but are not relevant to be put into Search class
+ */
+export default class MetadataManager {
```
[the remark this answers is not in the document]
Comment (lipim12) [reply src/api/MetadataManager.js]: In future (and other PRs) this class is utilized for metadata management - other than save search
